### PR TITLE
Document the security extra for SNI support with legacy Python

### PR DESCRIPTION
**Pull request by browniebroke**
This was added back in #2195.

This started as I was trying to understand why we had PyOpenSSL installed on the project I joined recently. After a bit of digging, it seems that it came from the time when we were using Python 2 and was for Requests/urllib3.

However, looking at Requests' documentation, I couldn't find a clear explanation about this feature, so this is my attempt at fixing this.

Disclaimer: I'm by no means an expert at SNI, so please correct me on anything wrong that follows.

Here are what I could find:

- The intend was to [document this feature](https://github.com/requests/requests/pull/2195#issuecomment-54524390) after merging it, so I assume this contribution is welcome.
- Looking at [the documentation from urllib3](https://urllib3.readthedocs.io/en/latest/user-guide.html#ssl-py2), it seems that the extra dependencies aren't needed in Python 3* and by the look of #4315 it can slow down importing Requests when we have PyOpenSSL.
- Another open pull request (#4591) is trying to change this, but @sigmavirus24 comment implies that there is more to SNI in Requests?

I'm not sure if this is the best place to document this, adding it to the "install" section feels overwhelming for beginners, and I felt like it belongs close to the SSL & certificates verification of the advanced section. Happy to move it somewhere else.

*Edit:*

(*) Actually, it seems it was implemented in [PEP 476](https://www.python.org/dev/peps/pep-0476/) which covers Python >=2.7.9 or >=[3.4.3](https://docs.python.org/3.4/whatsnew/changelog.html#python-3-4-3-final), so not any Python 3 versions.

### CODE REVIEW

File: docs/user/advanced.rst [hunk 1]
```diff
@@ -269,6 +269,28 @@ If you specify a wrong path or an invalid cert, you'll get a SSLError::
 .. warning:: The private key to your local certificate *must* be unencrypted.
    Currently, Requests does not support using encrypted keys.
 
+.. _sni-support:
+
+SNI support with legacy Python
+------------------------------
+
+.. versionadded:: 2.4.1
```
Comment (browniebroke) [inline docs/user/advanced.rst]: I've put this `versionadded` directive, but since it's quoting a very old version, it might not make sense?

Comment (sigmavirus24) [reply docs/user/advanced.rst]: This was originally added for SNI support but further is really just a way of providing modern OpenSSL for old versions of Python too. It allows them to communicate with modern websites regardless of SNI.

Comment (browniebroke) [reply docs/user/advanced.rst]: Thanks for your feedback. I've updated the change to be less focused on SNI, and direct to the upstream documentation for `urllib3`.

**Comment by sethmlarson** [comment]
Closing this as we're not recommending the security extra any longer.

**Pull request closed without merging**